base_layer/core/src/consensus/emission.rs

```diff
@@ -59,8 +54,17 @@ impl EmissionSchedule {
     /// $$ \epsilon = \sum 2^{-k} \foreach k \in decay $$
     ///
     /// So for example, if the decay rate is 0.25, then $$\epsilon$$ is 0.75 or 1/2 + 1/4 i.e. `1 >> 1 + 1 >> 2`
-    /// and the decay array is `&[1, 2]`
+    /// and the decay array is `&[1, 2]`.
+    ///
+    /// ## Panics
+    ///
+    /// The shift right operation will overflow if shifting more than 63 bits. `new` will panic if any of the decay
+    /// values are greater than or equal to 64.
     pub fn new(initial: MicroTari, decay: &'static [u64], tail: MicroTari) -> EmissionSchedule {
+        assert!(
+            decay.iter().all(|i| *i < 64),
+            "Decay value would overflow. All `decay` values must be less than 64"
+        );
```

Comment:
Won't this only work in debug/test code and not production?

Perhaps the function should be swapped to return a result rather?

Reply:
Too late this has already been merged into dev 🤣  (just appears in this diff because this is merging to tari-script)

Assert will panic in release too, however since the values for decay are hard coded (importantly, NOT dynamic) consensus values so i think it's ok to do this for the sake of keeping the function more ergonimic/creating an error just for the "programmer hard coded the wrong values" case